tools/traffic/table/blob_table.go

```diff
@@ -0,0 +1,125 @@
+)
+
+// BlobTable tracks blobs written by the traffic generator. This is a thread safe data structure.
+type BlobTable struct {
+
+	// blobs contains all blobs currently tracked by the requiredReads.
+	blobs []*BlobMetadata
+
+	// size describes the total number of blobs currently tracked by the requiredReads.
+	// size may be smaller than the capacity of the blobs slice.
+	size uint
+
+	// lock is used to synchronize access to the requiredReads.
+	lock sync.Mutex
+}
+
+// NewBlobTable creates a new BlobTable instance.
+func NewBlobTable() BlobTable {
+	return BlobTable{
+		blobs: make([]*BlobMetadata, 1024),
+		size:  0,
+	}
+}
+
+// Size returns the total number of blobs currently tracked by the requiredReads.
+func (table *BlobTable) Size() uint {
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	return table.size
+}
+
+// Get returns the blob at the specified index. Returns nil if the index is out of bounds.
+func (table *BlobTable) Get(index uint) *BlobMetadata {
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	if index >= table.size {
+		return nil
+	}
+
+	return table.blobs[index]
+}
+
+// Add a blob to the requiredReads.
+func (table *BlobTable) Add(blob *BlobMetadata) {
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	blob.index = table.size
+	table.blobs[table.size] = blob
+	table.size++
+}
+
+// AddOrReplace adds a blob to the requiredReads if there is capacity or replaces an existing blob at random
+// if the requiredReads is full. This method is a no-op if maximumCapacity is 0.
+func (table *BlobTable) AddOrReplace(blob *BlobMetadata, maximumCapacity uint) {
+	if maximumCapacity == 0 {
+		return
+	}
+
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	if table.size >= maximumCapacity {
+		// replace random existing blob
+		index := rand.Int31n(int32(table.size))
+		table.blobs[index] = blob
+		blob.index = uint(index)
+	} else {
+		// add new blob
+		blob.index = table.size
+		table.blobs[table.size] = blob
+		table.size++
+	}
+}
+
+// GetRandom returns a random blob currently tracked by the requiredReads. Returns nil if the requiredReads is empty.
+// Optionally decrements the read  permits of the blob if decrement is true. If the number of read permits
+// reaches 0, the blob is removed  from the requiredReads. Returns the blob metadata (if there is at least one blob
+// in the table) and a boolean indicating whether the blob was removed from the table as a result of this operation.
+func (table *BlobTable) GetRandom(decrement bool) (*BlobMetadata, bool) {
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	if table.size == 0 {
+		return nil, false
+	}
+
+	blob := table.blobs[rand.Int31n(int32(table.size))]
+
+	removed := false
+	if decrement && blob.remainingReadPermits != -1 {
+		blob.remainingReadPermits--
+		if blob.remainingReadPermits == 0 {
+			table.remove(blob)
+			removed = true
+		}
+	}
+
+	return blob, removed
+}
+
+// remove a blob from the requiredReads.
+func (table *BlobTable) remove(blob *BlobMetadata) {
```

Comment:
Is there a particular reason why we need table to be a slice?
I feel like removing/adding elements would be a lot simpler and less brittle if we used a map

Reply:
Primary reason why I use a slice is to give an `O(1)` implementation of `GetRandom()`. Can you think of a good way to do this backed by a map?

Reply:
Do we need to access a random element from the slice? 
For sampling, could we just use the first element from the map?

Reply:
The entire reason why this complex data structure exists in the first place is to facilitate random access. 😜 

For the blobs with a number of required reads: yes, we could get away without random access. But I'm under the impression a random access pattern is preferable to a fixed one when simulating workloads like this. 

For the pool of optional blobs to read, I think random access is necessary. Otherwise, we'd just be reading the same blob over and over until we get a new blob to start reading.

I'm open to discussing this more in depth if you are not convinced by my reasoning.

Reply:
> For the blobs with a number of required reads: yes, we could get away without random access. But I'm under the impression a random access pattern is preferable to a fixed one when simulating workloads like this.

We _could_ make get semi random access with a map in constant time (just generate random number n < 10 and pick _n_th element from map) if the access doesn't require sampling from uniform distribution. 

> For the pool of optional blobs to read, I think random access is necessary. Otherwise, we'd just be reading the same blob over and over until we get a new blob to start reading.

I don't think optional blob reads were ever part of the spec. The primary goal for this observability tool is to monitor if the network can handle a given retrieval traffic. Are there benefits of saturating the network with optional reads?

Not a big deal, but would bias toward simplicity vs. optimization

Reply:
Refactored to use a map based implementation.